Modify rule S5205,S5213: Cover std::move_only_function (CPP-5047) (#3775)
SonarSource/rspec: rules/S5205/cfamily/rule.adoc

== Why is this an issue?

> When you want to define a function that is parametrized by another function, there are three ways to declare its parameter type:

Reviewer: I believe the subject in the second part of the sentence is a bit unclear.

What about
```suggestion
When you want to receive a function pointer as a parameter in a function definition, there are three ways to declare its parameter type:
```

Author: I wanted to avoid "function pointer" in this sentence because it may not be what you want to use in the end. That's why I went for "parametrized by another function".

That said, I appreciate the sentence is not easy to get... Happy to get any other suggestions.

Reviewer: Okay, then a rephrase putting the parameter as the subject? Or just get rid of the "pointer" all together? (keeping the function as the subject, easier to read IMO)
```suggestion
When you want to receive a function as a parameter in a function definition, there are three ways to declare its parameter type:
```

Reviewer: Upon re-reading, this repeats "parameter", and without it, the same reference to the subject ("its") becomes ambiguous. I'm fine with either

Using a function pointer is an inferior solution for the following reasons:

* Only a function pointer can be passed as an argument, while the other options offer the caller more flexibility because they can take more advanced functors, such as lambdas with some captured state
* The syntax is obscure
* It typically has worse performance than the template parameter solution.

See S5213 for a discussion on choosing between ``++std::function++`` and a template parameter.

=== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
using Criterion = bool (*)(DataPoint const&);
void filter(DataSet& data, Criterion criterion); // Noncompliant

using Callback = void (*)(EventInfo const&);
class Button {
public:
  void addOnClick(Callback c) { // Noncompliant
    myOnClickHandler = c;
  }

private:
  Callback myOnClickHandler;
};
----

=== Compliant solution

[source,cpp,diff-id=1,diff-type=compliant]
----
template <class Criterion> // Compliant, uses the more efficient template argument
void filter(DataSet& data, Criterion criterion);

using Callback = std::function<void(EventInfo const&)>;
class Button {
public:

>   void addOnClick(Callback c) { // Compliant, uses the more flexible std::function
>     myOnClickHandler = c;
>   }

Reviewer: For when a std::function is passed as parameter
```suggestion
  void addOnClick(Callback const& c) { // Compliant, uses the more flexible std::function
    myOnClickHandler = c;
  }
```
or
```suggestion
  void addOnClick(Callback c) { // Compliant, uses the more flexible std::function
    myOnClickHandler = std::move(c);
  }
```
